x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -144,229 +139,331 @@ public void authorize(Authentication authentication, String action, TransportReq
-            }
-            if (permission.indices().check(action)) {
-                auditTrail.accessGranted(auditId, authentication, action, request, permission.names());
-                return;
+            final ActionListener<AuthorizationResult> clusterAuthzListener = wrapPreservingContext(ActionListener.wrap(result -> {
+                if (result.isGranted()) {
+                    if (result.isAuditable()) {
+                        auditTrail.accessGranted(requestId, authentication, action, unwrappedRequest, authzInfo);
+                    }
+                    putTransientIfNonExisting(AuthorizationServiceField.INDICES_PERMISSIONS_KEY, IndicesAccessControl.ALLOW_ALL);
+                    listener.onResponse(null);
+                } else {
+                    authzEngine.checkSameUserPermissions(authentication, unwrappedRequest, action, authzInfo,
+                        wrapPreservingContext(ActionListener.wrap(sameUserResult -> {
+                                if (sameUserResult.isGranted()) {
+                                    if (sameUserResult.isAuditable()) {
+                                        auditTrail.accessGranted(requestId, authentication, action, unwrappedRequest, authzInfo);
+                                    }
+                                    putTransientIfNonExisting(AuthorizationServiceField.INDICES_PERMISSIONS_KEY,
+                                        IndicesAccessControl.ALLOW_ALL);
+                                    listener.onResponse(null);
+                                } else {
+                                    listener.onFailure(denial(requestId, authentication, action, unwrappedRequest, authzInfo));
+                                }
+                            }, e -> listener.onFailure(denial(requestId, authentication, action, unwrappedRequest, authzInfo))),
+                            threadContext));
+                }
+            }, e -> {
+                // TODO need a failure handler better than this!
+                listener.onFailure(denial(requestId, authentication, action, unwrappedRequest, authzInfo, e));
+            }), threadContext);
+            authzEngine.authorizeClusterAction(authentication, unwrappedRequest, action, authzInfo, clusterAuthzListener);
+        } else if (IndexPrivilege.ACTION_MATCHER.test(action)) {
+            if (TransportActionProxy.isProxyAction(action)) {
+                // we've already validated that the request is a proxy request so we can skip that but we still
+                // need to validate that the action is allowed and then move on
+                authorizeIndexActionName(authentication, action, requestId, unwrappedRequest, authzInfo, authzEngine, listener);
+            } else if (authzEngine.shouldAuthorizeIndexActionNameOnly(action, unwrappedRequest)) {
+                authorizeIndexActionName(authentication, action, requestId, unwrappedRequest, authzInfo, authzEngine, listener);
```

Comment:
I'm not fundamentally opposed to this approach, but it seems strange, and I assume it's being done to try and optimise performance, but it creates an API that feels a bit weird.

Why ask the engine whether it wants to authorize by name? Why not just let it authorize however it wants - why does the service care?
I take it that we do this so we can avoid doing the index name resolution if it's not going to be used, but I think the API would be improved if we did by passing a `Supplier` or something like that to a single `authorizeIndexAction` method, and let the engine call it if/when it needed to.

Although I see from your comments below that some of the flow here is based on the desire to keep BWC in responses. If that's the case here, then perhaps we can come back to this as a master only change later, but I have a gut feel that if we push more of this flow control into the engine, then the BWC could  be handled by the RBAC engine rather than here.